Add script for setting up perf machines

DrewScoggins opened the pull request:
This script downloads and installs:
- Git
- CMake
- Java
- Python 2.7
- Python 3.6
- Visual Studio 2015 Community Edition

We also setup nuget.exe in the tools directory on the machine, and set
the power profiule to High Performance.

DrewScoggins committed:
Add script for setting up perf machines

This script downloads and installs:
Git
CMake
Java
Python 2.7
Python 3.6
Visual Studio 2015 Community Edition

We also setup nuget.exe in the tools directory on the machine, and set
the power profiule to High Performance.

DrewScoggins commented:
PTAL @jorive @dpodder @adiaaida

jorive commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -0,0 +1,144 @@
+﻿#Set power scheme to high performance
+powercfg /S 8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c
```
I would suggest replacing using the guid with this: `powercfg /setactive SCHEME_MIN`

jorive replied:
Should we disable hibernation too? `powercfg /hibernate off`

jorive replied:
These are other powercfg commands we should consider:
```
powercfg SUB_SLEEP        STANDBYIDLE     0  
powercfg SUB_VIDEO        VIDEOIDLE       0  
powercfg SUB_PROCESSOR    PERFBOOSTMODE   0  
powercfg SUB_PROCESSOR    PROCTHROTTLEMAX 100
powercfg SUB_PROCESSOR    PROCTHROTTLEMIN 100
powercfg SUB_PROCESSOR    SYSCOOLPOL      1  
powercfg SUB_PROCESSOR    THROTTLING      0  
```

jorive replied:
You might require, at the top of the script, running as administrator ~~by checking something like this:
`([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] "Administrator")`~~
by adding `#Requires -RunAsAdministrator` at the top of the script.

jorive replied:
I will suggest to check for $LASTEXITCODE, in every step, and exit script as soon as there is a failure.

dpodder replied:
I second all the extra suggestions, though some caution: the extra powercfg commands (`SUB_*`) modify whichever scheme is active. When we implemented similar functionality for the native compiler perf tests, we created a new scheme as a clone of `SCHEME_MIN` and modified that one; it might be worth doing that here as well, unless we know these machines are going to be used for perf only and don't mind leaving them in peak performance mode (incl. no sleep and video always on) 24/7.

DrewScoggins replied:
These machines should never be used for anything other than performance runs so I am not too worried about over-writing the default `SCHEME_MIN`

dpodder commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -0,0 +1,144 @@
+﻿#Set power scheme to high performance
```
Recommend the following powershell headers for better maintainability:
```powershell
#Requires -Version 4
[CmdletBinding()] param()
Set-StrictMode -Version Latest
$ErrorActionPreference = 'Stop'
```

dpodder replied:
($ErrorActionPreference may or may not be a good idea, depending on whether things we call write to stderr or not; but without it, we should pass a lot more error checking calls, like `-ErrorAction 'Stop'` to calls like mkdir, etc.)

DrewScoggins replied:
These seem like good hygiene things.  I will add them.

dpodder commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -0,0 +1,144 @@
+Start-Process -FilePath vs_community.exe -ArgumentList $vsArgList -Wait
+
+mkdir C:\Tools
+wget https://dist.nuget.org/win-x86-commandline/latest/nuget.exe -OutFile C:\Tools\nuget.exe
```
Have you considered using `chocolatey` to install the various 3rd party components? I think they have packages for everything short of vs_community.

DrewScoggins replied:
I had not.  I just put this together yesterday when I was making a new machine to add to the lab.  I will look into using `chocolatey` to grab all of the non-VS components.

dpodder commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -0,0 +1,144 @@
+
+wget https://benchviewstorage.blob.core.windows.net/public/vs_community.exe -OutFile vs_community.exe
+#region VS Admin File
+$vsAdminFile = @"
```
Could this be moved into a separate config file? It would help readability, plus better XML intellisense.

michellemcdaniel replied:
Second this comment.

DrewScoggins replied:
Sure.  My initial desire was to make this an all-in-one script that you could just put on a machine, run and it would setup the machines.  But since we would probably just bring it down off of github having additional config files that come down as well should not be that hard.

dpodder reviewed (commented):
Thanks @DrewScoggins, a few comments inline

michellemcdaniel commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -0,0 +1,144 @@
+wget https://www.python.org/ftp/python/3.6.0/python-3.6.0.exe -OutFile python36.exe
+$python36ArgList = "/quiet", "PrependPath=1", "TargetDir=C:\Python36"
+echo Start-Process -FilePath python36.exe -ArgumentList $python36ArgList -Wait
+Start-Process -FilePath python36.exe -ArgumentList $python36ArgList -Wait
```
This is for setting up new perf machines, right? From this, I am assuming that we are assuming python 3.6.0 as our python3 version for windows machines, but our scripts run on linux assume the machines have 3.5. Are we planning on upgrading our linux boxes to also use 3.6 (and therefore need to upgrade our scripting)?

DrewScoggins replied:
This is a good question.  I am not opposed to just locking onto Python 3.5 as that would require the least amount of work to existing assets.  I am not sure what we would be losing by not rolling forward.

DrewScoggins committed:
Change to Python3.5

DrewScoggins committed:
Automate more setup steps

DrewScoggins committed:
Update setup script

DrewScoggins committed:
Update git install to 2.14.1

jorive commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -27,10 +27,10 @@ EnableSymlinks=Disabled
 #endregion
 $gitInstallLog > gitInstallInfo.log
-wget https://github.com/git-for-windows/git/releases/download/v2.12.0.windows.1/Git-2.12.0-64-bit.exe -OutFile git-2.12.exe
+wget https://github.com/git-for-windows/git/releases/download/v2.14.1.windows.1/Git-2.14.1-64-bit.exe -OutFile git-2.14.1.exe
```
Could you please set the version to a variable (i.e. `$RequiredGitVersion='2.14.1'`) and use it where needed?

jorive replied:
Could you please replace `echo` with `Write-Host`? `echo` tokenizes the string on spaces and prints them one per line.

jorive commented:
We should be using [chocolatey](https://chocolatey.org/) to manage our packages:
```cmd
choco install VisualStudio2015Community
choco install git.install
```

jorive commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -0,0 +1,177 @@
+$gitInstallLog > gitInstallInfo.log
+wget https://github.com/git-for-windows/git/releases/download/v2.14.1.windows.1/Git-2.14.1-64-bit.exe -OutFile git-2.14.1.exe
+$gitArgList = "/LOADINF=gitInstallInfo.log", "/VERYSILENT"
+echo Start-Process -FilePath .\git-2.14.1.exe -ArgumentList $gitArgList -Wait
```
We should replace the `echo` commands with `Write-Host`. The first command takes an array of strings and prints the on a different line, the second command will print a single line.

DrewScoggins committed:
Update scheduled task generation

Change the scheduled task to be on start up instead of at at log in.
Also add a prompt to get that admin password for the machine being setup
as we need that in order to have the machines run the tests at the
correct permission level.

jorive commented:
Should we change the `C:\` to something like `(Get-Location).Drive.Root`

jorive commented on machineSetup/setupPerfMachine.ps1:
```diff
@@ -0,0 +1,178 @@
+echo Start-Process -FilePath python35.exe -ArgumentList $python35ArgList -Wait
+Start-Process -FilePath python35.exe -ArgumentList $python35ArgList -Wait
+
+wget https://benchviewstorage.blob.core.windows.net/public/vs_community.exe -OutFile vs_community.exe
```
We should use this Url: `https://download.microsoft.com/download/0/B/C/0BC321A4-013F-479C-84E6-4A2F90B11269/vs_community.exe` directly from Microsoft, or move to choco. The BenchView storage does not exist anymore.

jorive commented:
I do not know if there is another way, but we might need to run this at the beginning of the script:
```powershell
[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12
```
Otherwise, on a freshly installed Windows machine I get this on the `wget/Invoke-WebRequest` calls:
```log
Invoke-WebRequest : The request was aborted: Could not create SSL/TLS secure channel.
```
